### PR TITLE
feat: major upgrades of many packages and configuration

### PR DESCRIPTION
**Pull request by abutterworth**
This is a very large set of updates. We should verify everything is still configured as needed. The themeable and static builds still build as expected.

### CODE REVIEW

File: .npmignore [hunk 1]
```diff
@@ -1,12 +1,9 @@
 .DS_Store
-.eslintcache
 .storybook
-node_modules
```
Comment (matthugs) [inline .npmignore]: Why'd we add this back into the files sent to npm on publish?

Larger question I'm wondering about: can we switch to the whitelist approach described in link.medium.com/FcQryrm4bV, while we're in here mucking with which files get packed up?

File: .travis.yml [hunk 2]
```diff
@@ -1,15 +1,14 @@
...
-- 6
 cache:
   directories:
   - "~/.npm"
 notifications:
   email:
     recipients:
-    - arizzitano@edx.org
-    - jbradley@edx.org
+    - abutterworth@edx.org
```
Comment (matthugs) [inline .travis.yml]: we can fairly easily add a google group recipient here if you'd like to not be the only one feeling the spam (should spam arise)

File: package.json [hunk 3]
```diff
@@ -2,21 +2,22 @@
...
   "scripts": {
-    "build": "NODE_ENV=production webpack --config webpack.config.js",
+    "build": "webpack --mode production",
     "build-storybook": "build-storybook",
+    "commit": "commit",
     "coveralls": "cat ./coverage/lcov.info | coveralls",
     "debug-test": "node --inspect-brk node_modules/.bin/jest --runInBand --coverage",
     "deploy-storybook": "storybook-to-ghpages",
     "deploy-storybook-ci": "storybook-to-ghpages --ci",
     "gc": "commit",
```
Comment (matthugs) [inline package.json]: can we get rid of this alternative alias? This name I find very counterintuitive

File: package.json [hunk 4]
```diff
@@ -26,80 +27,71 @@
...
+    "source-map-loader": "^0.2.4",
+    "storybook-readme": "^4.0.5",
+    "style-loader": "^0.23.1",
+    "travis-deploy-once": "^5.0.11",
+    "uglifyjs-webpack-plugin": "^2.1.2",
+    "webpack": "^4.29.6",
+    "webpack-cli": "^3.3.0"
   },
   "jest": {
+    "testURL": "http://localhost/",
```
Comment (matthugs) [inline package.json]: this appears to be the default https://jestjs.io/docs/en/configuration.html#testurl-string

File: .storybook/addons.js [hunk 5]
```diff
@@ -1,4 +1,2 @@
-import '@storybook/addon-options/register';
 import '@storybook/addon-actions/register';
 import '@storybook/addon-a11y/register';
```
Comment (matthugs) [inline .storybook/addons.js]: It looks like we removed all the places the corresponding decorator is used. Should we strip this out too, or are we planning on adding them back in piecemeal?

File: src/ListBox/ListBox.stories.jsx [hunk 6]
```diff
@@ -145,7 +142,7 @@ storiesOf('ListBox', module)
     </ListBox>
   ))
   .add('using tag prop', () => (
-    <React.Fragment>
+    <div>
```
Comment (matthugs) [inline src/ListBox/ListBox.stories.jsx]: nit: why?

File: src/ListBox/index.jsx [hunk 7]
```diff
@@ -104,9 +103,9 @@ export default class ListBox extends React.Component {
 }
 
 ListBox.propTypes = {
-  children: or([
-    elementType(ListBoxOption),
-    PropTypes.arrayOf(elementType(ListBoxOption)),
+  children: PropTypes.oneOfType([
+    PropTypes.arrayOf(PropTypes.element),
+    PropTypes.element,
```
Comment (matthugs) [inline src/ListBox/index.jsx]: Consider also: `PropTypes.node`

File: src/Modal/index.jsx [hunk 8]
```diff
@@ -139,21 +139,17 @@ class Modal extends React.Component {
...
-      buttonElement = (<Button
-        {...buttonProps}
-        key={i}
-        onKeyDown={this.handleKeyDown}
-      />);
-
-      return buttonElement;
+    return this.props.buttons.map((button) => {
+      // button is either a Button component that we want clone or a set of props
+      const buttonProps = button.type === Button ? button.props : button;
```
Comment (matthugs) [inline src/Modal/index.jsx]: Why clone at all? I feel like this adds the kind of react hierarchy-indirection that makes me nervous after ADR-2 in this repo https://github.com/edx/paragon/blob/master/docs/decisions/0002-react-i18n.rst

File: .travis.yml (hunk 2 above)
Comment (abutterworth) [reply .travis.yml]: Good call I'll make that happen

File: .storybook/addons.js (hunk 5 above)
Comment (abutterworth) [reply .storybook/addons.js]: The new version of this addon doesn't need to be added with addDecorator – it seems this line is all you need to get set up.

File: src/Modal/index.jsx (hunk 8 above)
Comment (abutterworth) [reply src/Modal/index.jsx]: It seems that this was made to handle these two different ways of providing buttons. I agree we shouldn't do what we are here, but for the sake of backward compatibility I didn't want to change anything. This was more just cleaning up some code that I found hard to read.

File: .npmignore (hunk 1 above)
Comment (abutterworth) [reply .npmignore]: Great! I replaced this with the whitelist

File: package.json (hunk 4 above)
Comment (abutterworth) [reply package.json]: I think I added this due to an issue with Storyshots. I can remove and see if it breaks anything.

File: package.json (hunk 3 above)
Comment (abutterworth) [reply package.json]: removed.

File: src/ListBox/index.jsx (hunk 7 above)
Comment (abutterworth) [reply src/ListBox/index.jsx]: Thanks that's better.

File: src/ListBox/ListBox.stories.jsx (hunk 6 above)
Comment (abutterworth) [reply src/ListBox/ListBox.stories.jsx]: Good catch – no good reason.

File: src/Modal/index.jsx (hunk 8 above)
Comment (matthugs) [reply src/Modal/index.jsx]: My jaw would drop if any downstream users of this component relied on the behavior of us not using the same buttons they pass in, but prop-clones of the same.

But I can understand the desire to push off this concern; should I make a note of it to drop in the backlog once we have one set up?

File: package.json (hunk 3 above)
Comment (matthugs) [reply package.json]: Sorry I was unclear about which line I was pointing to in github's interface; it's much clear when adding a comment

Your name makes loads more sense to me than `npm run gc`. It's a pretty bad false-cognate with `git gc`

File: src/Modal/index.jsx (hunk 8 above)
Comment (abutterworth) [reply src/Modal/index.jsx]: Yes please do.

File: package.json (hunk 4 above)
Comment (abutterworth) [reply package.json]: No problems removing it 👍

File: package.json (hunk 3 above)
Comment (abutterworth) [reply package.json]: Got it, changed.

File: src/Modal/index.jsx (hunk 8 above)
Comment (matthugs) [reply src/Modal/index.jsx]: This ended up being mooted in https://github.com/edx/paragon/pull/435 by the elimination of these buttons from the modal component's prop interface. Marking as resolved.

**Pull request merged**